### DISCUSSION TITLE
Questions on `nbd_do_recv()`

### DISCUSSION THREAD

**Discussion by derekbit** [ming1/ubdsrv#48] [General]
Hello experts,

I'm studying the implementation of the ublk-nbd's [nbd_do_recv()](https://github.com/ming1/ubdsrv/blob/master/nbd/tgt_nbd.cpp#LL434C9-L434C9). There are some questions arising when I was reading the codes.

1. For `const int loops = len < 512 ? 16 : 32;`,  why does it set `loops` to a smaller number when `len` is smaller than `32` instead of `32` for any len?
2. The receive logic does `recv()` first and then `nbd_start_recv` (io_uring method)? Why not just use `recv()` or `nbd_start_recv (io_uring method)`? Does hybrid approach have any advantage?
3. What's the purpose of the `fake_cqe` in [`__nbd_handle_recv()`](https://github.com/ming1/ubdsrv/blob/master/nbd/tgt_nbd.cpp#L468)

Many thanks for helps in advance.

**Comment by ming1** [comment]
> Hello experts,
> 
> I'm studying the implementation of the ublk-nbd's [nbd_do_recv()](https://github.com/ming1/ubdsrv/blob/master/nbd/tgt_nbd.cpp#LL434C9-L434C9). There are some questions arising when I was reading the codes.
> 
>     1. For `const int loops = len < 512 ? 16 : 32;`,  why does it set `loops` to a smaller number when `len` is smaller than `32` instead of `32` for any len?
> 

len < 512 means len is 16, which is for receiving reply, so probably 16 is 
enough for polling reply, and 32 is for polling data.

>     2. The receive logic does `recv()` first and then `nbd_start_recv` (io_uring method)? Why not just use `recv()` or `nbd_start_recv (io_uring method)`? Does hybrid approach have any advantage?
> 

It improves whole ublk-nbd much, and you can observe performance change by killing it via simply setting loops as zero(by taking completely io_uring method).

Just use recv() may block the current context for long time, so won't work.
Using complete io_uring method may add a bit latency for recv,
since there may be lots of send requests queued.

This is really good question, and I think there is big chance to improve current implementation. Any feedback or patches are welcome.

>     3. What's the purpose of the `fake_cqe` in [`__nbd_handle_recv()`](https://github.com/ming1/ubdsrv/blob/master/nbd/tgt_nbd.cpp#L468)
> 

It is just to pass data and wake up __nbd_handle_io_async(), which needs
to check io->tgt_io_cqe->res. The idea is that nbd write request handling
result is got from nbd reply message.

Thanks,
Ming
